infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -143,75 +154,46 @@
 
     /**
      * Constructor Discovery Service
-     *
      * @param serverContext current server's context
      * @param clusterManagerAdapter adapter to communicate to external Cluster Manager
      * @param serverCallback callback to Log Replication Server upon discovery
-     *
      */
     public CorfuReplicationDiscoveryService(@Nonnull ServerContext serverContext,
                                             @Nonnull CorfuReplicationClusterManagerAdapter clusterManagerAdapter,
                                             @Nonnull CompletableFuture<CorfuInterClusterReplicationServerNode> serverCallback) {
+        this.executorService = Executors.newSingleThreadExecutor(new
+                ThreadFactoryBuilder().setNameFormat("discovery-service-executor").build());
         this.clusterManagerAdapter = clusterManagerAdapter;
         this.logReplicationNodeId = serverContext.getNodeId();
         this.serverContext = serverContext;
         this.localEndpoint = serverContext.getLocalEndpoint();
         this.serverCallback = serverCallback;
+        this.shutdown = false;
+        this.eventQueue = new LinkedBlockingQueue<>();
+        this.executorService.submit(this::run);
```

Comment:
I think it is good to initialize below 'isLeader' before we call run/start.

Reply:
+1